### PR TITLE
myriad/service_config

### PR DESCRIPTION
**Pull request by eyadof**

### CODE REVIEW

File: lib/Myriad.pm [hunk 1]
```diff
@@ -628,6 +629,8 @@ async method run () {
         }))
     }
 
+    $self->storage;
```
Comment (tom-binary) [inline lib/Myriad.pm:632]: This looks wrong - what's it doing? If it's setting up the initial storage, would expect `await` or at least a comment.

File: lib/Myriad/API.pm [hunk 2]
```diff
@@ -60,6 +63,21 @@ method service_by_name ($name) {
...
+
+=head2 config
+
+Returns a L<Ryu::Observable> that hold the value of the configuration.
+
+=cut
+
+method config ($key) {
+    my $pkg = caller;
+    if($Myriad::Config::SERVICES_CONFIG{$pkg}->{$key}) {
```
Comment (tom-binary) [inline lib/Myriad/API.pm:75]: This is wrong - config is per-instance

File: lib/Myriad/Config.pm [hunk 3]
```diff
@@ -60,61 +71,42 @@ The C<< %SHORTCUTS_FOR >> hash allows commandline shortcuts for common parameter
 =cut
 
 our %SHORTCUTS_FOR = (
```
Comment (tom-binary) [inline lib/Myriad/Config.pm:73]: the name is now backwards!

File: lib/Myriad/Config.pm [hunk 4]
```diff
@@ -135,11 +129,131 @@ method define ($key, $v) {
...
+    die 'config key is not formated correctly' unless @sublist;
+    while (@sublist > 1) {
+        my $level = shift @sublist;
+        $root->{$level} //= {};
+        $root= $root->{$level};
+    }
+    $root->{$sublist[0]} = $value;
+}
+
+method from_args ($commandline) {
```
Comment (tom-binary) [inline lib/Myriad/Config.pm:150]: Moving these out is good, but they need documentation and the names aren't really following the right convention: each of these *does* something, so it needs a verb. `->from_args` implies that it's returning data - `my $something = $obj->from_something_else`.

File: lib/Myriad/Config.pm [hunk 5]
```diff
@@ -135,11 +129,131 @@ method define ($key, $v) {
...
+    if ($error) {
+        $log->error($error);
+        die pod2usage(1);
+    }
+}
+
+method from_env () {
+    $config->{$_} //= delete $ENV{'MYRIAD_' . uc($_)} for grep { exists $ENV{'MYRIAD_' . uc($_)} } keys %DEFAULTS;
+    map {
+        $_ =~ s/(MYRIAD_SERVICES?_)//;
```
Comment (tom-binary) [inline lib/Myriad/Config.pm:191]: The `$_ =~` is implicit, can leave it out

File: lib/Myriad.pm (hunk 1 above)
Comment (eyadof) [reply lib/Myriad.pm:632]: yes indeed, this a master hack fixing this case is what I'm working on next, but for historical accuracy, things are going to be handled in the `on_startup` processing so no need to `await` this is basically to just instantiate the object

File: lib/Myriad/API.pm (hunk 2 above)
Comment (eyadof) [reply lib/Myriad/API.pm:75]: yes and that is handled in `Myriad::Config` this is not to look up for the config value this is just to check if the developer declared this before or not.

**Pull request merged**